x-pack/filebeat/input/etw/input.go

```diff
@@ -0,0 +1,296 @@
+		return fmt.Errorf("error initializing ETW session: %w", err)
+	}
+
+	// Set up logger with session information
+	e.log = ctx.Logger.With("session", e.etwSession.Name)
+	e.log.Info("Starting " + inputName + " input")
+
+	// Handle realtime session creation or attachment
+	if e.etwSession.Realtime {
+		if !e.etwSession.NewSession {
+			// Attach to an existing session
+			err = e.operator.AttachToExistingSession(e.etwSession)
+			if err != nil {
+				return fmt.Errorf("unable to retrieve handler: %w", err)
+			}
+			e.log.Debug("attached to existing session")
+		} else {
+			// Create a new realtime session
+			err = e.operator.CreateRealtimeSession(e.etwSession)
+			if err != nil {
+				return fmt.Errorf("realtime session could not be created: %w", err)
+			}
+			e.log.Debug("created session")
+		}
+	}
+	// Defer the cleanup and closing of resources
+	var wg sync.WaitGroup
+	var once sync.Once
+
+	// Create an error channel to communicate errors from the goroutine
+	errChan := make(chan error, 1)
+
+	defer func() {
+		once.Do(e.Close)
+		e.log.Info(inputName + " input stopped")
+	}()
+
+	// eventReceivedCallback processes each ETW event
+	eventReceivedCallback := func(record *etw.EventRecord) uintptr {
+		if record == nil {
+			e.log.Error("received null event record")
+			return 1
+		}
+
+		e.log.Debugf("received event %d with length %d", record.EventHeader.EventDescriptor.Id, record.UserDataLength)
+
+		data, err := etw.GetEventProperties(record)
+		if err != nil {
+			e.log.Errorf("failed to read event properties: %w", err)
+			return 1
+		}
+
+		evt := beat.Event{
+			Timestamp: time.Now(),
+			Fields: mapstr.M{
+				"metadata": fillEventMetadata(e.etwSession, e.config),
+				"header":   fillEventHeader(record.EventHeader),
+				"winlog":   data,
+			},
+		}
+
+		publisher.Publish(evt)
+
+		return 0
+	}
+
+	// Set the callback function for the ETW session
+	e.etwSession.Callback = eventReceivedCallback
+
+	// Start a goroutine to consume ETW events
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		e.log.Debug("starting to listen ETW events")
+		if err = e.operator.StartConsumer(e.etwSession); err != nil {
+			errChan <- fmt.Errorf("failed to start consumer: %w", err) // Send error to channel
+			return
+		}
+		e.log.Debug("stopped to read ETW events from session")
+		errChan <- nil
+	}()
+
+	// We ensure resources are closed when receiving a cancelation signal
+	go func() {
+		<-ctx.Cancelation.Done()
+		once.Do(e.Close)
+	}()
+
+	wg.Wait() // Ensure all goroutines have finished before closing
+	close(errChan)
+	if err, ok := <-errChan; ok && err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// fillEventHeader constructs a header map for an event record header.
+func fillEventHeader(h etw.EventHeader) map[string]interface{} {
+	// Mapping from Level to Severity
+	levelToSeverity := map[uint8]string{
+		1: "critical",
+		2: "error",
+		3: "warning",
+		4: "information",
+		5: "verbose",
+	}
+
+	header := make(map[string]interface{})
```

Comment:
It appears that you know what the keys will be in this map. So you can initialize and allocate the map at the same time. The size is known so it can allocate sufficient space (which should be more efficient).

```go
header := map[string]any{
    "size": h.Size,
   "type": h.HeaderType,
}
```